File: src/logger.js
const { config } = require('./config');

const LogLevel = {
  ERROR: 0,
  WARN: 1,
  INFO: 2,
  DEBUG: 3,
};

const levelMap = {
  error: LogLevel.ERROR,
  warn: LogLevel.WARN,
  info: LogLevel.INFO,
  debug: LogLevel.DEBUG,
};

const currentLevel = levelMap[config.logging.level] || LogLevel.INFO;

const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m',
  gray: '\x1b[90m',
};

function getTimestamp() {
  const now = new Date();
  return now.toISOString();
}

function log(level, message, data = null) {
  if (levelMap[level] > currentLevel) return;

  const timestamp = getTimestamp();
  const levelColors = {
    error: colors.red,
    warn: colors.yellow,
    info: colors.cyan,
    debug: colors.gray,
  };

  const colorizedLevel = `${levelColors[level]}${level.toUpperCase()}${colors.reset}`;
  const output = `[${timestamp}] ${colorizedLevel} ${message}`;

  if (level === 'error') {
    console.error(output);
    if (data) console.error(data);
  } else {
    console.log(output);
    if (data) console.log(data);
  }
}

module.exports = {
  info: (msg, data) => log('info', msg, data),
  error: (msg, data) => log('error', msg, data),
  warn: (msg, data) => log('warn', msg, data),
  debug: (msg, data) => log('debug', msg, data),
};

File: logger.py
import logging
import sys
from config import Config

# Map string log levels to logging constants
LOG_LEVEL_MAP = {
    'DEBUG': logging.DEBUG,
    'INFO': logging.INFO,
    'WARNING': logging.WARNING,
    'ERROR': logging.ERROR,
    'CRITICAL': logging.CRITICAL
}

class CustomFormatter(logging.Formatter):
    """Logging Formatter to add colors and count warning / errors"""

    grey = "\x1b[38;21m"
    cyan = "\x1b[36m"
    yellow = "\x1b[33;21m"
    red = "\x1b[31;21m"
    bold_red = "\x1b[31;1m"
    reset = "\x1b[0m"
    format_str = "[%(asctime)s] %(levelname)s %(message)s"

    FORMATS = {
        logging.DEBUG: grey + format_str + reset,
        logging.INFO: cyan + format_str + reset,
        logging.WARNING: yellow + format_str + reset,
        logging.ERROR: red + format_str + reset,
        logging.CRITICAL: bold_red + format_str + reset
    }

    def format(self, record):
        log_fmt = self.FORMATS.get(record.levelno)
        formatter = logging.Formatter(log_fmt, datefmt='%Y-%m-%d %H:%M:%S')
        return formatter.format(record)

def setup_logger():
    logger = logging.getLogger('AmaranthBot')
    
    # Set level
    level = LOG_LEVEL_MAP.get(Config.LOG_LEVEL, logging.INFO)
    logger.setLevel(level)

    # Create console handler
    # Force UTF-8 for console output to avoid cp949 errors on Windows
    if sys.platform == 'win32':
        try:
            sys.stdout.reconfigure(encoding='utf-8', errors='replace')
        except Exception:
            pass
        
    ch = logging.StreamHandler(sys.stdout)
    ch.setLevel(level)

    # Set formatter
    ch.setFormatter(CustomFormatter())

    # Add handler to logger
    if not logger.handlers:
        logger.addHandler(ch)

    return logger

logger = setup_logger()
